pkg/services/ngalert/state/manager_bench_test.go

```diff
@@ -0,0 +1,106 @@
+package state_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+	"github.com/grafana/grafana/pkg/services/ngalert/state/historian"
+	"github.com/stretchr/testify/mock"
+)
+
+func BenchmarkProcessEvalResults(b *testing.B) {
+	as := annotations.FakeAnnotationsRepo{}
+	as.On("SaveMany", mock.Anything, mock.Anything).Return(nil)
+	hist := historian.NewAnnotationBackend(&as, nil, nil)
+	cfg := state.ManagerCfg{
+		Historian: hist,
+	}
+	sut := state.NewManager(cfg)
+	now := time.Now().UTC()
+	rule := makeBenchRule()
+	results := makeBenchResults(100)
+	labels := map[string]string{}
+
+	var ans []state.StateTransition
+	for i := 0; i < b.N; i++ {
+		ans = sut.ProcessEvalResults(context.Background(), now, &rule, results, labels)
+	}
+
+	b.StopTimer()
+
+	_ = fmt.Sprintf("%v", len(ans))
+}
+
+func makeBenchRule() models.AlertRule {
+	dashUID := "my-dash"
+	panelID := int64(14)
+	return models.AlertRule{
+		ID:              5,
+		OrgID:           1,
+		Title:           "some rule",
+		Condition:       "A",
+		Data:            []models.AlertQuery{},
```

Comment:
Another place where I left some fields out - rules are overall a small portion of what the system operates on, so this doesn't seem to make much difference in the actual results. We can extend this if reviewers would prefer, though.